src/gpgpu-sim/local_interconnect.cc

```diff
@@ -206,41 +204,47 @@ void xbar_router::iSLIP_Advance() {
     cycles_util++;
   }
   // do iSLIP
-  for (unsigned i = 0; i < total_nodes; ++i) {
-    if (Has_Buffer_Out(i, 1)) {
-      for (unsigned j = 0; j < total_nodes; ++j) {
-        unsigned node_id = (j + next_node[i]) % total_nodes;
-
-        if (!in_buffers[node_id].empty()) {
+  for (auto dest : destination_set) {
+    if (Has_Buffer_Out(dest, 1)) {
+      unsigned start_node = next_node[dest];
+      auto it = std::upper_bound(input_nodes.begin(), input_nodes.end(),
+                                       start_node);
+      for (unsigned j = 0; j < input_nodes.size(); j++, it++) {
+        if (it == input_nodes.end()) {
+          it = input_nodes.begin();
+        }
+        unsigned node_id = *it;
```

Comment:
I have a concern whether this piece of code is equivalent to the original code.

I understand that `std::set<unsigned> destination_set` is a subset of the range from 0 to total_nodes. The outer loop iterates over it. So, `i` is equivalent to `dest`.

In the original code, the first value which `node_id` can be assigned is `(0 + next_node[i]) % total_nodes`.

In your PR, the first value `node_id` can be assigned is `(1 + next_node[dest]) % total_nodes`. This offset of 1 comes from `std::upper_bound` which yields the first iterator `iter` that can satisfy `bool(comp(start_node, *iter))==true`. (citing cppreference.com). `comp(start_node, start_node)` yields `false`.

Which one is correct? Or does this matter?

Reply:
The point of this is round-robin. So starting from 0 or 1 should be fine either way. The difference should be minimal.

Reply:
Is this specified in the ISLIP paper or not? If it was, better do it right.

Reply:
if it wasn't specified in the paper, I'm fine with the PR